File: apps/cli/src/commands/timeline.ts
import { hasFlag, json, valueAfter } from "../cli-shared.js";
import { formatSnapError } from "@qxinm/gandalf-core/errors.js";
import { findTimelineEntry, listTimelineEntries, type TimelineCorruptEvent } from "@qxinm/gandalf-core/store.js";
import { buildTimelineUndoPlan } from "@qxinm/gandalf-core/timeline-undo.js";
import type { TimelineEntry } from "@qxinm/gandalf-core/types.js";
import type { Command, CommandContext } from "./index.js";

export const timelineCommand: Command = {
  name: "timeline",
  description: "List and show local timeline entries",

  async execute(ctx: CommandContext): Promise<number> {
    const { args, options } = ctx;
    const sub = args[1] ?? "list";

    if (sub === "list") {
      const corruptEvents: TimelineCorruptEvent[] = [];
      const entries = await listTimelineEntries(options.storeDir, {
        agent: options.agent,
        projectPath: options.projectPath,
        limit: limitFromArgs(args),
        onCorruptEntry: (event) => corruptEvents.push(event)
      });
      reportCorruptEvents(corruptEvents);

      if (hasFlag(args, "--json")) {
        process.stdout.write(json(entries));
        return 0;
      }

      process.stdout.write(renderTimelineList(entries));
      return 0;
    }

    if (sub === "show") {
      const ref = args[2];
      if (!ref) {
        process.stderr.write(
          formatSnapError({
            code: "GANDALF_TIMELINE_REF_REQUIRED",
            problem: "Timeline entry id or snapshot name is required.",
            cause: "`timeline show` was called without a reference.",
            fix: "Run `gandalf timeline list` and pass an entry id or snapshot name."
          })
        );
        return 1;
      }

      const corruptEvents: TimelineCorruptEvent[] = [];
      const entry = await findTimelineEntry(options.storeDir, ref, {
        onCorruptEntry: (event) => corruptEvents.push(event)
      });
      reportCorruptEvents(corruptEvents);
      if (!entry) {
        process.stderr.write(
          formatSnapError({
            code: "GANDALF_TIMELINE_ENTRY_NOT_FOUND",
            problem: `Timeline entry "${ref}" not found.`,
            cause: "The reference does not match a timeline id or snapshot name.",
            fix: "Run `gandalf timeline list` to see available entries."
          })
        );
        return 1;
      }

      process.stdout.write(hasFlag(args, "--json") ? json(entry) : renderTimelineEntry(entry));
      return 0;
    }

    if (sub === "undo") {
      const ref = args[2];
      if (!ref) {
        process.stderr.write(
          formatSnapError({
            code: "GANDALF_TIMELINE_REF_REQUIRED",
            problem: "Timeline entry id or snapshot name is required.",
            cause: "`timeline undo` was called without a reference.",
            fix: "Run `gandalf timeline list` and pass an entry id."
          })
        );
        return 1;
      }
      if (!hasFlag(args, "--dry-run")) {
        process.stderr.write(
          formatSnapError({
            code: "GANDALF_TIMELINE_UNDO_DRY_RUN_REQUIRED",
            problem: "Timeline undo is dry-run only in P0.",
            cause: "`timeline undo` was called without `--dry-run`.",
            fix: "Run `gandalf timeline undo <id> --dry-run --json`."
          })
        );
        return 1;
      }

      const corruptEvents: TimelineCorruptEvent[] = [];
      let plan: Awaited<ReturnType<typeof buildTimelineUndoPlan>>;
      try {
        plan = await buildTimelineUndoPlan(options.storeDir, ref, {
          onCorruptEntry: (event) => corruptEvents.push(event)
        });
      } catch (error) {
        reportCorruptEvents(corruptEvents);
        process.stderr.write(
          formatSnapError({
            code: "GANDALF_TIMELINE_ENTRY_NOT_FOUND",
            problem: `Timeline entry "${ref}" not found.`,
            cause: error instanceof Error ? error.message : "The reference does not match a readable timeline id or snapshot name.",
            fix: "Run `gandalf timeline list` to see available entries and any corrupt event warnings."
          })
        );
        return 1;
      }
      reportCorruptEvents(corruptEvents);
      process.stdout.write(hasFlag(args, "--json") ? json(plan) : renderTimelineUndoPlan(plan));
      return 0;
    }

    process.stderr.write(
      formatSnapError({
        code: "GANDALF_UNKNOWN_SUBCOMMAND",
        problem: `Unknown timeline subcommand: "${sub}".`,
        cause: "`timeline` was called with an unrecognized subcommand.",
        fix: "Use `list`, `show`, or `undo`. Run `gandalf --help` for details."
      })
    );
    return 1;
  }
};

function limitFromArgs(args: string[]): number | undefined {
  const raw = valueAfter(args, "--limit");
  if (!raw) return undefined;
  const parsed = Number.parseInt(raw, 10);
  return Number.isFinite(parsed) && parsed >= 0 ? parsed : undefined;
}

function renderTimelineList(entries: TimelineEntry[]): string {
  if (entries.length === 0) {
    return "No timeline entries.\n";
  }

  const lines = ["gandalf timeline", ""];
  for (const entry of entries) {
    const scope = entry.agent ? ` ${entry.agent}` : "";
    lines.push(`${entry.id}  ${entry.observedAt}  ${entry.restoreReadiness}${scope}  ${entry.title}`);
  }
  return `${lines.join("\n")}\n`;
}

function renderTimelineEntry(entry: TimelineEntry): string {
  const lines = [
    entry.title,
    "",
    `id: ${entry.id}`,
    `event: ${entry.eventKind}`,
    `after snapshot: ${entry.afterSnapshotName}`,
    `before snapshot: ${entry.beforeSnapshotName ?? "-"}`,
    `observed: ${entry.observedAt}`,
    `source: ${entry.source}`,
    `capture: ${entry.captureId}`,
    `project: ${entry.projectPath}`,
    `agent: ${entry.agent ?? "all"}`,
    `restore readiness: ${entry.restoreReadiness}`,
    `confidence: ${entry.confidence} (${entry.confidenceReason})`,
    "",
    "Summary",
    `  evidence: ${entry.evidenceCount}`,
    `  graph nodes: ${entry.graphNodeCount}`,
    `  audit findings: ${entry.auditFindingCount}`,
    `  semantic changes: ${entry.changes.semanticChangeCount}`,
    `  raw source changes: ${entry.changes.rawSourceChangeCount}`
  ];

  if (entry.changes.previousSnapshotName) {
    lines.push(`  previous snapshot: ${entry.changes.previousSnapshotName}`);
  }

  if (entry.changes.highlights.length > 0) {
    lines.push("", "Highlights");
    for (const highlight of entry.changes.highlights) {
      lines.push(`  ${highlight}`);
    }
  }

  return `${lines.join("\n")}\n`;
}

function renderTimelineUndoPlan(plan: Awaited<ReturnType<typeof buildTimelineUndoPlan>>): string {
  const lines = [
    plan.title,
    "",
    `entry: ${plan.entryId}`,
    `dry run: yes`,
    `writes files: no`,
    `restore readiness: ${plan.restoreReadiness}`,
    "",
    "MCP changes"
  ];

  if (plan.writableItems.length === 0) {
    lines.push("  none");
  } else {
    for (const item of plan.writableItems) {
      lines.push(`  ${item.action} ${item.serverName} in ${item.path}`);
    }
  }

  if (plan.observeOnlySurfaces.length > 0) {
    lines.push("", "Observe-only surfaces");
    for (const surface of plan.observeOnlySurfaces) {
      lines.push(`  ${surface.changeType}: ${surface.entityName ?? surface.path}`);
    }
  }

  return `${lines.join("\n")}\n`;
}

function reportCorruptEvents(events: TimelineCorruptEvent[]): void {
  for (const event of events) {
    process.stderr.write(`Skipped corrupt timeline event: ${event.filePath} (${event.error})\n`);
  }
}

export default timelineCommand;

File: internal/cli/timeline.go
package cli

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/qyinm/hem/internal/hemcore/store"
	"github.com/qyinm/hem/internal/hemcore/types"
	"github.com/spf13/cobra"
)

func newTimelineCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "timeline",
		Short: "List and inspect local timeline entries",
	}
	cmd.AddCommand(newTimelineListCmd())
	cmd.AddCommand(newTimelineShowCmd())
	cmd.AddCommand(newTimelineUndoCmd())
	return cmd
}

func newTimelineListCmd() *cobra.Command {
	var common CommonFlags

	cmd := &cobra.Command{
		Use:   "list",
		Short: "List timeline entries",
		RunE: func(cmd *cobra.Command, _ []string) error {
			exitCode := runTimelineList(cmd, &common)
			if exitCode != 0 {
				return errExit(exitCode)
			}
			return nil
		},
	}

	common.bindFlags(cmd.Flags())
	return cmd
}

func newTimelineShowCmd() *cobra.Command {
	var common CommonFlags

	cmd := &cobra.Command{
		Use:   "show [reference]",
		Short: "Show a timeline entry by id or snapshot name",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			exitCode := runTimelineShow(cmd, &common, args[0])
			if exitCode != 0 {
				return errExit(exitCode)
			}
			return nil
		},
	}

	common.bindFlags(cmd.Flags())
	return cmd
}

func newTimelineUndoCmd() *cobra.Command {
	var common CommonFlags
	var dryRun bool

	cmd := &cobra.Command{
		Use:   "undo [reference]",
		Short: "Build a dry-run MCP undo plan for a timeline entry",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			exitCode := runTimelineUndo(cmd, &common, args[0], dryRun)
			if exitCode != 0 {
				return errExit(exitCode)
			}
			return nil
		},
	}

	common.bindFlags(cmd.Flags())
	cmd.Flags().BoolVar(&dryRun, "dry-run", false, "Preview undo plan without making changes")
	return cmd
}

func runTimelineList(cmd *cobra.Command, common *CommonFlags) int {
	runtime, snapErr := resolveRuntime(common)
	if snapErr != nil {
		return writeError(cmd.ErrOrStderr(), snapErr)
	}

	var corruptEvents []store.TimelineCorruptEvent
	entries, err := store.ListTimelineEntries(runtime.StoreDir, store.TimelineListOptions{
		Agent:       runtime.Agent,
		ProjectPath: runtime.ProjectPath,
		OnCorruptEntry: func(event store.TimelineCorruptEvent) {
			corruptEvents = append(corruptEvents, event)
		},
	})
	if err != nil {
		return writeError(cmd.ErrOrStderr(), &types.SnapError{
			Code:    "HEM_TIMELINE_LIST_FAILED",
			Problem: "Failed to list timeline entries.",
			Cause:   err.Error(),
			Fix:     "Verify the store directory is readable.",
		})
	}

	for _, event := range corruptEvents {
		_, _ = fmt.Fprintf(cmd.ErrOrStderr(), "Skipped corrupt timeline event: %s (%s)\n", event.FilePath, event.Error)
	}

	if common.JSON {
		return writeJSON(cmd.OutOrStdout(), entries)
	}
	if len(entries) == 0 {
		return writeStdout(cmd.OutOrStdout(), "No timeline entries.\n")
	}

	lines := []string{"hem timeline", ""}
	for _, entry := range entries {
		lines = append(lines, fmt.Sprintf(
			"- %s %s (%s) -> %s",
			entry.ID,
			entry.Title,
			entry.EventKind,
			entry.AfterSnapshotName,
		))
	}
	return writeStdout(cmd.OutOrStdout(), strings.Join(lines, "\n")+"\n")
}

func runTimelineShow(cmd *cobra.Command, common *CommonFlags, reference string) int {
	runtime, snapErr := resolveRuntime(common)
	if snapErr != nil {
		return writeError(cmd.ErrOrStderr(), snapErr)
	}

	var corruptEvents []store.TimelineCorruptEvent
	entry, err := store.FindTimelineEntry(runtime.StoreDir, reference, store.TimelineListOptions{
		Agent:       runtime.Agent,
		ProjectPath: runtime.ProjectPath,
		OnCorruptEntry: func(event store.TimelineCorruptEvent) {
			corruptEvents = append(corruptEvents, event)
		},
	})
	if err != nil {
		return writeError(cmd.ErrOrStderr(), &types.SnapError{
			Code:    "HEM_TIMELINE_LOOKUP_FAILED",
			Problem: "Failed to look up timeline entry.",
			Cause:   err.Error(),
			Fix:     "Run `hem timeline list` to see available entries.",
		})
	}
	for _, event := range corruptEvents {
		_, _ = fmt.Fprintf(cmd.ErrOrStderr(), "Skipped corrupt timeline event: %s (%s)\n", event.FilePath, event.Error)
	}
	if entry == nil {
		return writeError(cmd.ErrOrStderr(), &types.SnapError{
			Code:    "HEM_TIMELINE_NOT_FOUND",
			Problem: fmt.Sprintf("Timeline entry not found: %q.", reference),
			Cause:   "The reference does not match a timeline id or snapshot name.",
			Fix:     "Run `hem timeline list` to see available entries.",
		})
	}

	if common.JSON {
		return writeJSON(cmd.OutOrStdout(), entry)
	}
	data, err := json.MarshalIndent(entry, "", "  ")
	if err != nil {
		return writeError(cmd.ErrOrStderr(), &types.SnapError{
			Code:    "HEM_JSON_SERIALIZE_FAILED",
			Problem: "Failed to serialize timeline entry.",
			Cause:   err.Error(),
			Fix:     "This is an internal error.",
		})
	}
	return writeStdout(cmd.OutOrStdout(), string(data)+"\n")
}

func runTimelineUndo(cmd *cobra.Command, _ *CommonFlags, _ string, dryRun bool) int {
	if !dryRun {
		return writeError(cmd.ErrOrStderr(), &types.SnapError{
			Code:    "HEM_TIMELINE_UNDO_DRY_RUN_REQUIRED",
			Problem: "Timeline undo requires --dry-run.",
			Cause:   "`timeline undo` was called without `--dry-run`.",
			Fix:     "Run `hem timeline undo <id> --dry-run --json`.",
		})
	}
	return writeError(cmd.ErrOrStderr(), notImplementedError("hem timeline undo"))
}